# Youtube Music Analysis Project Part 2: Music Analysis

This notebook explores and analyses my YouTube watch history using Python and pandas. this notebook will be performing exploratory data analysis with a focus on music listening habits

## Analysis Goals

Youtube Recap already records the following metrics (which this notebook will also attempt to record):

- Most listened to songs/Artists/Genres/Musical Moods
- Total Listening Time

However, in my project I will also attempt to record, analyse and visualise other statistics due to the Google Takeout data containing exact timestamps of when a video was viewed/listened, we can use this to analyse:
- Listening Habits - When exactly do I listen to music
- Listening Sessions - How long/How many songs do I listen to in a given session
- Song loyalty - How many times do I listen to the same song? How often do i listen to new songs?
- Song trends - How many times do I listen to songs over time

What do I listen to?
Songs, artists, genres, favourites, total listening time.

When do I listen?
Hour, day, weekday/weekend, sessions, heatmaps.

How do I listen?
Repeats, diversity, session length, concentration.

How does my listening change?
Discovery, trends, obsessions, fading songs, changing artists/genres.

## Basic Analysis

Retrieving:
1. Total number of music plays
2. Number of unique videos listened to
3. Number of unique channels

In [11]:
from string.templatelib import convert

import pandas as pd

music_dataframe = pd.read_csv(r"C:\Users\MsnSi\PycharmProjects\Youtube_Music_Analysis\data\processed\music_history.csv")

music_dataframe.head()

,title,channel,watched_at,video_url,video_ids,channel_id,video_duration,classification,classification_reason,classification_evidence
0,Midnight Grand Orchestra『Moonlightspeed 』Music...,Midnight Grand Orchestra,"1 Aug 2026, 11:47:44 IST",https://www.youtube.com/watch?v=NI8KqzCvFsc,NI8KqzCvFsc,UCnVbtCwr-5LXxUlGxsgD7sQ,PT3M25S,confirmed_music,music_channel,NaN
1,Midnight Grand Orchestra『Moonlightspeed 』Music...,Midnight Grand Orchestra,"1 Aug 2026, 11:47:44 IST",https://www.youtube.com/watch?v=NI8KqzCvFsc,NI8KqzCvFsc,UCnVbtCwr-5LXxUlGxsgD7sQ,PT3M25S,confirmed_music,music_channel,NaN
2,Midnight Grand Orchestra『Moonlightspeed 』Music...,Midnight Grand Orchestra,"1 Aug 2026, 11:47:44 IST",https://www.youtube.com/watch?v=NI8KqzCvFsc,NI8KqzCvFsc,UCnVbtCwr-5LXxUlGxsgD7sQ,PT3M25S,confirmed_music,music_channel,NaN
3,Midnight Grand Orchestra『Moonlightspeed 』Music...,Midnight Grand Orchestra,"1 Aug 2026, 11:47:44 IST",https://www.youtube.com/watch?v=NI8KqzCvFsc,NI8KqzCvFsc,UCnVbtCwr-5LXxUlGxsgD7sQ,PT3M25S,confirmed_music,music_channel,NaN
4,「3時12分 / TAKU INOUE & 星街すいせい」MUSIC VIDEO,TAKU INOUE,"1 Aug 2026, 11:54:27 IST",https://www.youtube.com/watch?v=LYFciXBcXIQ,LYFciXBcXIQ,UCICy3BC4CIU2OXa7VGoektw,PT3M13S,confirmed_music,music_channel,NaN


In [12]:
music_plays = len(music_dataframe)
unique_videos = music_dataframe["video_ids"].nunique()
unique_channels = music_dataframe["channel_id"].nunique()

print(f"In the DataFrame, there are:\n"
      f"{music_plays} number of plays\n"
      f"{unique_videos} unique videos\n"
      f"{unique_channels} unique channels\n")

In the DataFrame, there are:
3097 number of plays
279 unique videos
168 unique channels



Next, we'll retrieve the most commonly played videos and channels

In [13]:
video_counts = music_dataframe["video_ids"].value_counts()
channel_counts = music_dataframe["channel"].value_counts()
title_counts = music_dataframe["title"].value_counts()

print(title_counts.head(10))
print(channel_counts.head(10))


title
Nightcore / Sped Up → Ai Đưa Em Về (Rock Version/Cover)                      484
Why Do I (with Hatsune Miku)                                                 441
奏  - スキマスイッチ (Cover) / VESPERBELL ヨミ                                         256
Imagine Dragons - Take Me To The Beach (feat. Ado) (Official Lyric Video)    196
アンノウン・マザーグース（cover）/ 相沢                                                      144
【歌ってみた】喜劇 カバー 羊羊｜フルVer. #星野源 #SPY_FAMILY                                     144
Ms. VICTORIA                                                                 121
Umapyoi Celebrity                                                             81
【Ado】オールナイトレディオ                                                               64
I'm a chud song (Kasane Teto Cover)                                           64
Name: count, dtype: int64
channel
Sen                                 490
Set It Off                          468
VESPERBELL_YM                       256
ImagineDragonsVEVO            

Interestingly, the results surprised me slightly, that Why Do I, wasn't the most frequently played song as was recorded in my actual June - August Youtube Recap, although of course, this Data set is different, going from the 7th of July to the 7th of September,

In [14]:
top_1_plays = title_counts.head(1).sum()
top_3_plays = title_counts.head(3).sum()
top_5_plays = title_counts.head(5).sum()
top_10_plays = title_counts.head(10).sum()
top_25_plays = title_counts.head(25).sum()

top_1_plays_percentage = (top_1_plays / music_plays) * 100
top_3_plays_percentage = (top_3_plays / music_plays) * 100
top_5_plays_percentage = (top_5_plays / music_plays) * 100
top_10_plays_percentage = (top_10_plays / music_plays) * 100
top_25_plays_percentage = (top_25_plays / music_plays) * 100

print(f"The Top 1 Song has {top_1_plays} plays with {top_1_plays_percentage:.1f}% of plays\n"
      f"The Top 3 Songs have {top_3_plays} plays with {top_3_plays_percentage:.1f}% of plays\n"
      f"The Top 5 Songs have {top_5_plays} plays with {top_5_plays_percentage:.1f}% of plays\n"
      f"The Top 10 Songs have {top_10_plays} plays with {top_10_plays_percentage:.1f}% of plays\n"
      f"The Top 25 Songs have {top_25_plays} plays with {top_25_plays_percentage:.1f}% of plays\n")

The Top 1 Song has 484 plays with 15.6% of plays
The Top 3 Songs have 1181 plays with 38.1% of plays
The Top 5 Songs have 1521 plays with 49.1% of plays
The Top 10 Songs have 1995 plays with 64.4% of plays
The Top 25 Songs have 2461 plays with 79.5% of plays



Interestingly, the most played songs take up a very large amount of the listening amount


In [15]:
music_dataframe[["title", "video_duration"]].head()

,title,video_duration
0,Midnight Grand Orchestra『Moonlightspeed 』Music...,PT3M25S
1,Midnight Grand Orchestra『Moonlightspeed 』Music...,PT3M25S
2,Midnight Grand Orchestra『Moonlightspeed 』Music...,PT3M25S
3,Midnight Grand Orchestra『Moonlightspeed 』Music...,PT3M25S
4,「3時12分 / TAKU INOUE & 星街すいせい」MUSIC VIDEO,PT3M13S


## Analysing Listening Habits

Firstly before anything else, one of the pieces of data within our dataset is the duration of the video, but it is currently in the form PTxHyMzS which we'll need to convert into an actual duration if we want to use it

In [16]:
import re
duration_1 = "PT4M52S"
duration_2 = "PT1H2M14S"
duration_3 = "PT47S"
duration_4 = "PT5M"

durations = [duration_1, duration_2, duration_3, duration_4]


def convert_ISO8601_to_string(duration):
    match = re.match(
        r"PT(?:(\d+)H)?(?:(\d+)M)?(?:(\d+)S)?",
        duration)

    return match.groups()

def convert_video_duration_to_seconds(duration):
    ISO_converted_string = convert_ISO8601_to_string(duration)

    seconds = 0
    x = 0
    multiplers = [3600, 60, 1]

    for x in range(3):
        if ISO_converted_string[x] is None:
            continue
        else:
            seconds += int(ISO_converted_string[x]) * multiplers[x]

    return seconds

for duration in durations:
    print(f"Converted {duration} into {convert_video_duration_to_seconds(duration)} seconds\n")


Converted PT4M52S into 292 seconds

Converted PT1H2M14S into 3734 seconds

Converted PT47S into 47 seconds

Converted PT5M into 300 seconds



In [17]:
music_dataframe["duration_seconds"] = (
    music_dataframe["video_duration"].apply(convert_video_duration_to_seconds)
)

print(music_dataframe["duration_seconds"].describe())

print(
    music_dataframe[
        ["title", "video_duration", "duration_seconds"]
    ].head(10)
)

count    3097.000000
mean      308.775266
std       775.822480
min        10.000000
25%       168.000000
50%       215.000000
75%       274.000000
max      7311.000000
Name: duration_seconds, dtype: float64
                                               title video_duration  \
0  Midnight Grand Orchestra『Moonlightspeed 』Music...        PT3M25S   
1  Midnight Grand Orchestra『Moonlightspeed 』Music...        PT3M25S   
2  Midnight Grand Orchestra『Moonlightspeed 』Music...        PT3M25S   
3  Midnight Grand Orchestra『Moonlightspeed 』Music...        PT3M25S   
4           「3時12分 / TAKU INOUE & 星街すいせい」MUSIC VIDEO        PT3M13S   
5       Ghost City Tokyo ♡ English Cover【rachie】幽霊東京         PT4M4S   
6       Ghost City Tokyo ♡ English Cover【rachie】幽霊東京         PT4M4S   
7                    Overdose - Kobo Kanaeru (Cover)        PT3M14S   
8                    Overdose - Kobo Kanaeru (Cover)        PT3M14S   
9                    Overdose - Kobo Kanaeru (Cover)        PT3M14S   

   duration

In [24]:
def convert_seconds_to_HMS(seconds):
    minutes, seconds = divmod(seconds, 60)
    hours, minutes = divmod(minutes, 60)
    return f"{hours:02d}H:{minutes:02d}M:{seconds:02d}S"


total_listening_seconds = music_dataframe["duration_seconds"].sum()
total_listening_time = convert_seconds_to_HMS(total_listening_seconds)

print(f"Total listening time in seconds: {total_listening_seconds}s\n"
      f"Total listening time: {total_listening_time}")


Total listening time in seconds: 956277s
Total listening time: 265H:37M:57S


265 Hours is quite a lot of time for listening, its likley we have some kind of outliers.

In [37]:
music_dataframe[
    ["title", "video_duration", "duration_seconds"]
].sort_values("duration_seconds", ascending=False).head(10)

,title,video_duration,duration_seconds
2739,Prison Labor - Persona 5 Royal OST [Extended],PT2H1M51S,7311
2740,Prison Labor - Persona 5 Royal OST [Extended],PT2H1M51S,7311
2162,Prison Labor - Persona 5 Royal OST [Extended],PT2H1M51S,7311
2163,Prison Labor - Persona 5 Royal OST [Extended],PT2H1M51S,7311
2164,Prison Labor - Persona 5 Royal OST [Extended],PT2H1M51S,7311
2165,Prison Labor - Persona 5 Royal OST [Extended],PT2H1M51S,7311
2166,Prison Labor - Persona 5 Royal OST [Extended],PT2H1M51S,7311
2167,Prison Labor - Persona 5 Royal OST [Extended],PT2H1M51S,7311
2735,Prison Labor - Persona 5 Royal OST [Extended],PT2H1M51S,7311
2736,Prison Labor - Persona 5 Royal OST [Extended],PT2H1M51S,7311


Yep, this video is about 2 hours long, but my last listen only ended about 10 minutes in, we'll need to fix this

In [51]:
long_videos = music_dataframe[
    music_dataframe["duration_seconds"] > 600
]

print(long_videos["title"].value_counts())

long_video_seconds = long_videos["duration_seconds"].sum()

print(convert_seconds_to_HMS(long_video_seconds))

title
Prison Labor - Persona 5 Royal OST [Extended]                                                       36
ヨルシカをメドレーにして1人でバンドしてみた -Yorushika Medley One man Band Cover                                          4
that one dinky ass kirby song except it loops                                                        1
星街すいせいメドレーを一人でバンド組んで歌ってみた                                                                            1
【アコギ】ヨルシカ Acoustic Guitarメドレー2【全12曲】                                                                 1
POV: you're in love with a memory | Playlist                                                         1
Mobile Suit Gundam Iron Blooded Orphans Urdr Hunt OST - Red vs Blue [Extended]                       1
At Bantam 🍷✨ Yakuza Bar/Cafe OST Mix                                                                 1
Pokemon Black and White - Low HP Music EXTENDED                                                      1
57 minutes of silence occasionally interrupted by weird route cue  

Because YouTube Watch History records that a video was watched but not
the exact watch duration, total listening time cannot be measured exactly.

The raw estimate assumes every music video was watched in full.

For this personal dataset, I made a note of the music tracks that were noted in this code and manually adjusted their duration to my recollection These
adjustments are therefore specific to this dataset and are not intended as a general listening-time algorithm.

In [58]:
duration_exceptions = {
    "Prison Labor - Persona 5 Royal OST [Extended]": 10 * 60,
    "that one dinky ass kirby song except it loops": 12 * 60,
    "POV: you're in love with a memory | Playlist": 2 * 60,
    "Mobile Suit Gundam Iron Blooded Orphans Urdr Hunt OST - Red vs Blue [Extended]": 2 * 60,
    "At Bantam 🍷✨ Yakuza Bar/Cafe OST Mix": 5 * 60,
    "Pokemon Black and White - Low HP Music EXTENDED": 2 * 60,
    "57 minutes of silence occasionally interrupted by weird route cue": 2 * 60,
    "Go To Sleep, She Isn't Coming Back.": 2 * 60,
    "Coldplay – Viva La Vida Loop on Cat Piano ($1000 request)": 2 * 60,
    "【YOASOBI🌷】全有名曲メドレー ｜Study with me｜Work with me｜リラックスタイム｜安眠・癒し・作業用BGM｜勉強用BGM｜AnnieChan_playlist ⑤": 10 * 60
}

for title, adjusted_duration in duration_exceptions.items():
    music_dataframe.loc[
        music_dataframe["title"] == title,
        "adjusted_duration_seconds"
    ] = adjusted_duration

total_adjusted_seconds = music_dataframe["adjusted_duration_seconds"].sum()

total_adjusted_listening_time = convert_seconds_to_HMS(
    total_adjusted_seconds
)

print(f"Raw estimated listening time: "
      f"{convert_seconds_to_HMS(total_listening_seconds)}")

print(f"Adjusted estimated listening time: "
      f"{total_adjusted_listening_time}")

Raw estimated listening time: 265H:37M:57S
Adjusted estimated listening time: 193H:06M:19S


,title,duration_seconds,adjusted_duration_seconds
772,POV: you're in love with a memory | Playlist,2034,120
